Does `crates/ruff/resources/test/fixtures/pyupgrade/UP014.py` contain incorrect `NamedTuple` calls?

<!--
Thank you for taking the time to report an issue! We're glad to have you involved with Ruff.

If you're filing a bug report, please consider including the following information:

* A minimal code snippet that reproduces the bug.
* The command you invoked (e.g., `ruff /path/to/file.py --fix`), ideally including the `--isolated` flag.
* The current Ruff settings (any relevant sections from your `pyproject.toml`).
* The current Ruff version (`ruff --version`).
-->

https://github.com/astral-sh/ruff/blob/f012ed2d77272bfdd75f6988c2658e36fbf63102/crates/ruff/resources/test/fixtures/pyupgrade/UP014.py#L7-L12

Is this a valid `NamedTuple` call?


```
> python3 --version
Python 3.11.4

> cat a.py
from typing import NamedTuple

MyType = NamedTuple(
    "MyType",
    [("a", int), ("b", str), ("c", list[bool])],
    defaults=["foo", [True]],
)

> python a.py
Traceback (most recent call last):
  File "/home/haru/Desktop/repositories/ruff/a.py", line 3, in <module>
    MyType = NamedTuple(
             ^^^^^^^^^^^
  File "/home/haru/miniconda3/envs/py311/lib/python3.11/typing.py", line 2916, in NamedTuple
    raise TypeError("Either list of fields or keywords"
TypeError: Either list of fields or keywords can be provided to NamedTuple, not both
```

> or is this intended?

Is this what you saw?

```py
from typing import NamedTuple
 
MyType = NamedTuple(
    "MyType",
    [("a", int), ("b", str), ("c", list[bool])],
    defaults=["foo", [True]],
)
```
```
crates/ruff_python_formatter/scratch.py:6:16: F821 Undefined name `foo`
Found 1 error.
```

--- a/crates/ruff/resources/test/fixtures/pyupgrade/UP014.py
+++ b/crates/ruff/resources/test/fixtures/pyupgrade/UP014.py
@@ -8,7 +8,6 @@ MyType = NamedTuple("MyType", [("a", int), ("b", tuple[str, ...])])
 MyType = NamedTuple(
     "MyType",
     [("a", int), ("b", str), ("c", list[bool])],
-    defaults=["foo", [True]],
 )
 
 # with namespace
@@ -18,7 +17,6 @@ MyType = typing.NamedTuple("MyType", [("a", int), ("b", str)])
 MyType = NamedTuple(
     "MyType",
     [("a", int), ("b", str)],
-    defaults=[1, "bar", "baz"],
 )
 
 # invalid identifiers (OK)
@@ -29,3 +27,6 @@ MyType = typing.NamedTuple("MyType")
 
 # empty fields
 MyType = typing.NamedTuple("MyType", [])
+
+# keywords
+MyType = typing.NamedTuple("MyType", a=int, b=tuple[str, ...])

--- a/crates/ruff/resources/test/fixtures/pyupgrade/UP014.py
+++ b/crates/ruff/resources/test/fixtures/pyupgrade/UP014.py
@@ -4,20 +4,8 @@ import typing
 # with complex annotations
 MyType = NamedTuple("MyType", [("a", int), ("b", tuple[str, ...])])
 
-# with default values as list
-MyType = NamedTuple(
-    "MyType",
-    [("a", int), ("b", str), ("c", list[bool])],
-)
-
 # with namespace
 MyType = typing.NamedTuple("MyType", [("a", int), ("b", str)])
-
-# too many default values (OK)
-MyType = NamedTuple(
-    "MyType",
-    [("a", int), ("b", str)],
-)
 
 # invalid identifiers (OK)
 MyType = NamedTuple("MyType", [("x-y", int), ("b", tuple[str, ...])])
@@ -30,3 +18,7 @@ MyType = typing.NamedTuple("MyType", [])
 
 # keywords
 MyType = typing.NamedTuple("MyType", a=int, b=tuple[str, ...])
+
+# unfixable
+MyType = typing.NamedTuple("MyType", [("a", int)], [("b", str)])
+MyType = typing.NamedTuple("MyType", [("a", int)], b=str)

--- a/crates/ruff/src/rules/pyupgrade/rules/convert_named_tuple_functional_to_class.rs
+++ b/crates/ruff/src/rules/pyupgrade/rules/convert_named_tuple_functional_to_class.rs
@@ -111,7 +111,7 @@ fn create_property_assignment_stmt(property: &str, annotation: &Expr) -> Stmt {
     .into()
 }
 
-/// Create a list of property assignments from the `NamedTuple` positional arguments.
+/// Create a list of property assignments from the `NamedTuple` fields argument.
 fn create_properties_from_fields_arg(fields: &Expr) -> Result<Vec<Stmt>> {
     let Expr::List(ast::ExprList { elts, .. }) = &fields else {
         bail!("Expected argument to be `Expr::List`");

--- a/crates/ruff/src/rules/pyupgrade/rules/convert_named_tuple_functional_to_class.rs
+++ b/crates/ruff/src/rules/pyupgrade/rules/convert_named_tuple_functional_to_class.rs
@@ -93,11 +93,7 @@ fn match_named_tuple_assign<'a>(
 
 /// Generate a `Stmt::AnnAssign` representing the provided property
 /// definition.
-fn create_property_assignment_stmt(
-    property: &str,
-    annotation: &Expr,
-    value: Option<&Expr>,
-) -> Stmt {
+fn create_property_assignment_stmt(property: &str, annotation: &Expr) -> Stmt {
     ast::StmtAnnAssign {
         target: Box::new(
             ast::ExprName {
@@ -108,34 +104,15 @@ fn create_property_assignment_stmt(
             .into(),
         ),
         annotation: Box::new(annotation.clone()),
-        value: value.map(|value| Box::new(value.clone())),
+        value: None,
         simple: true,
         range: TextRange::default(),
     }
     .into()
 }
 
-/// Match the `defaults` keyword in a `NamedTuple(...)` call.
-fn match_defaults(keywords: &[Keyword]) -> Result<&[Expr]> {
-    let defaults = keywords.iter().find(|keyword| {
-        if let Some(arg) = &keyword.arg {
-            arg == "defaults"
-        } else {
-            false
-        }
-    });
-    match defaults {
-        Some(defaults) => match &defaults.value {
-            Expr::List(ast::ExprList { elts, .. }) => Ok(elts),
-            Expr::Tuple(ast::ExprTuple { elts, .. }) => Ok(elts),
-            _ => bail!("Expected defaults to be `Expr::List` | `Expr::Tuple`"),
-        },
-        None => Ok(&[]),
-    }
-}
-
 /// Create a list of property assignments from the `NamedTuple` arguments.
-fn create_properties_from_args(args: &[Expr], defaults: &[Expr]) -> Result<Vec<Stmt>> {
+fn create_properties_from_args(args: &[Expr]) -> Result<Vec<Stmt>> {
     let Some(fields) = args.get(1) else {
         let node = Stmt::Pass(ast::StmtPass {
             range: TextRange::default(),
@@ -151,16 +128,8 @@ fn create_properties_from_args(args: &[Expr], defaults: &[Expr]) -> Result<Vec<S
         });
         return Ok(vec![node]);
     }
-    let padded_defaults = if elts.len() >= defaults.len() {
-        std::iter::repeat(None)
-            .take(elts.len() - defaults.len())
-            .chain(defaults.iter().map(Some))
-    } else {
-        bail!("Defaults must be `None` or an iterable of at least the number of fields")
-    };
     elts.iter()
-        .zip(padded_defaults)
-        .map(|(field, default)| {
+        .map(|field| {
             let Expr::Tuple(ast::ExprTuple { elts, .. }) = &field else {
                 bail!("Expected `field` to be `Expr::Tuple`")
             };
@@ -180,9 +149,7 @@ fn create_properties_from_args(args: &[Expr], defaults: &[Expr]) -> Result<Vec<S
             if is_dunder(property) {
                 bail!("Cannot use dunder property name: {}", property)
             }
-            Ok(create_property_assignment_stmt(
-                property, annotation, default,
-            ))
+            Ok(create_property_assignment_stmt(property, annotation))
         })
         .collect()
 }
@@ -228,9 +195,19 @@ pub(crate) fn convert_named_tuple_functional_to_class(
         return;
     };
 
-    let properties = match match_defaults(keywords)
-        .and_then(|defaults| create_properties_from_args(args, defaults))
-    {
+    // Ignore `NamedTuple` calls containing both a list of fields and keywords:
+    // ```
+    // NamedTuple(
+    //     "MyType",
+    //     [("a", int), ("b", str)],
+    //     defaults=[0, ""],
+    // )
+    // ```
+    if args.len() > 1 && !keywords.is_empty() {
+        return;
+    }
+
+    let properties = match create_properties_from_args(args) {
         Ok(properties) => properties,
         Err(err) => {
             debug!("Skipping `NamedTuple` \"{typename}\": {err}");

--- a/crates/ruff/src/rules/pyupgrade/rules/convert_named_tuple_functional_to_class.rs
+++ b/crates/ruff/src/rules/pyupgrade/rules/convert_named_tuple_functional_to_class.rs
@@ -155,8 +155,8 @@ fn create_properties_from_keywords(keywords: &[Keyword]) -> Result<Vec<Stmt>> {
         .map(|keyword| {
             let Keyword { arg, value, .. } = keyword;
             let Some(arg) = arg else {
-                    bail!("Expected `keyword` to have an `arg`")
-                };
+                bail!("Expected `keyword` to have an `arg`")
+            };
             Ok(create_property_assignment_stmt(arg.as_str(), value))
         })
         .collect()
@@ -205,29 +205,34 @@ pub(crate) fn convert_named_tuple_functional_to_class(
 
     let properties = match (&args[1..], keywords) {
         // Ex) NamedTuple("MyType")
-        ([], []) => {
-            let node = Stmt::Pass(ast::StmtPass {
-                range: TextRange::default(),
-            });
-            Ok(vec![node])
-        }
+        ([], []) => vec![Stmt::Pass(ast::StmtPass {
+            range: TextRange::default(),
+        })],
         // Ex) NamedTuple("MyType", [("a", int), ("b", str)])
-        ([fields], []) => create_properties_from_fields_arg(fields),
+        ([fields], []) => {
+            if let Ok(properties) = create_properties_from_fields_arg(fields) {
+                properties
+            } else {
+                debug!("Skipping `NamedTuple` \"{typename}\": unable to parse fields");
+                return;
+            }
+        }
         // Ex) NamedTuple("MyType", a=int, b=str)
-        ([], keywords) => create_properties_from_keywords(keywords),
+        ([], keywords) => {
+            if let Ok(properties) = create_properties_from_keywords(keywords) {
+                properties
+            } else {
+                debug!("Skipping `NamedTuple` \"{typename}\": unable to parse keywords");
+                return;
+            }
+        }
         // Unfixable
         _ => {
-            debug!("Skipping `NamedTuple` \"{typename}\": unfixable");
+            debug!("Skipping `NamedTuple` \"{typename}\": mixed fields and keywords");
             return;
         }
     };
-    let properties = match properties {
-        Ok(properties) => properties,
-        Err(err) => {
-            debug!("Skipping `NamedTuple` \"{typename}\": {err}");
-            return;
-        }
-    };
+
     let mut diagnostic = Diagnostic::new(
         ConvertNamedTupleFunctionalToClass {
             name: typename.to_string(),

--- a/crates/ruff/src/rules/pyupgrade/rules/convert_named_tuple_functional_to_class.rs
+++ b/crates/ruff/src/rules/pyupgrade/rules/convert_named_tuple_functional_to_class.rs
@@ -111,14 +111,8 @@ fn create_property_assignment_stmt(property: &str, annotation: &Expr) -> Stmt {
     .into()
 }
 
-/// Create a list of property assignments from the `NamedTuple` arguments.
-fn create_properties_from_args(args: &[Expr]) -> Result<Vec<Stmt>> {
-    let Some(fields) = args.get(1) else {
-        let node = Stmt::Pass(ast::StmtPass {
-            range: TextRange::default(),
-        });
-        return Ok(vec![node]);
-    };
+/// Create a list of property assignments from the `NamedTuple` positional arguments.
+fn create_properties_from_fields_arg(fields: &Expr) -> Result<Vec<Stmt>> {
     let Expr::List(ast::ExprList { elts, .. }) = &fields else {
         bail!("Expected argument to be `Expr::List`");
     };
@@ -150,6 +144,20 @@ fn create_properties_from_args(args: &[Expr]) -> Result<Vec<Stmt>> {
                 bail!("Cannot use dunder property name: {}", property)
             }
             Ok(create_property_assignment_stmt(property, annotation))
+        })
+        .collect()
+}
+
+/// Create a list of property assignments from the `NamedTuple` keyword arguments.
+fn create_properties_from_keywords(keywords: &[Keyword]) -> Result<Vec<Stmt>> {
+    keywords
+        .iter()
+        .map(|keyword| {
+            let Keyword { arg, value, .. } = keyword;
+            let Some(arg) = arg else {
+                    bail!("Expected `keyword` to have an `arg`")
+                };
+            Ok(create_property_assignment_stmt(arg.as_str(), value))
         })
         .collect()
 }
@@ -195,26 +203,31 @@ pub(crate) fn convert_named_tuple_functional_to_class(
         return;
     };
 
-    // Ignore `NamedTuple` calls containing both a list of fields and keywords:
-    // ```
-    // NamedTuple(
-    //     "MyType",
-    //     [("a", int), ("b", str)],
-    //     defaults=[0, ""],
-    // )
-    // ```
-    if args.len() > 1 && !keywords.is_empty() {
-        return;
-    }
-
-    let properties = match create_properties_from_args(args) {
+    let properties = match (&args[1..], keywords) {
+        // Ex) NamedTuple("MyType")
+        ([], []) => {
+            let node = Stmt::Pass(ast::StmtPass {
+                range: TextRange::default(),
+            });
+            Ok(vec![node])
+        }
+        // Ex) NamedTuple("MyType", [("a", int), ("b", str)])
+        ([fields], []) => create_properties_from_fields_arg(fields),
+        // Ex) NamedTuple("MyType", a=int, b=str)
+        ([], keywords) => create_properties_from_keywords(keywords),
+        // Unfixable
+        _ => {
+            debug!("Skipping `NamedTuple` \"{typename}\": unfixable");
+            return;
+        }
+    };
+    let properties = match properties {
         Ok(properties) => properties,
         Err(err) => {
             debug!("Skipping `NamedTuple` \"{typename}\": {err}");
             return;
         }
     };
-
     let mut diagnostic = Diagnostic::new(
         ConvertNamedTupleFunctionalToClass {
             name: typename.to_string(),

